modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersAuthenticator.java

```diff
@@ -0,0 +1,149 @@
+        final BasicHttpRequest requestToAuthenticate = wrapToBasicHttpRequest(request);
+        final Supplier<ThreadContext.StoredContext> emptyContext = threadContext.wrapRestorable(threadContext.newStoredContext());
+        final ActionListener<Void> contextPreservingListener = new ContextPreservingActionListener<>(emptyContext, listener);
+        try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+            populateThreadContext.apply(requestToAuthenticate, channel, threadContext);
+            authenticate.accept(requestToAuthenticate, ActionListener.wrap(ignored -> {
+                final ThreadContext.StoredContext authenticatedContext = threadContext.newStoredContext();
+                ((HttpHeadersWithAuthenticationContext) request.headers()).markAuthenticationSucceeded(authenticatedContext);
+                contextPreservingListener.onResponse(null);
+            }, e -> {
+                ((HttpHeadersWithAuthenticationContext) request.headers()).markAuthenticationFailed(e);
+                contextPreservingListener.onFailure(e);
+            }));
+        }
+    }
+
+    private static BasicHttpRequest wrapToBasicHttpRequest(HttpRequest request) {
+        return new BasicHttpRequest() {
+
+            @Override
+            public RestRequest.Method method() {
+                return Netty4HttpRequest.translateRequestMethod(request.method());
+            }
+
+            @Override
+            public String uri() {
+                return request.uri();
+            }
+
+            @Override
+            public Map<String, List<String>> getHeaders() {
+                return Netty4HttpRequest.wrapHttpHeaders(request.headers());
+            }
+        };
+    }
+
+    public static ThreadContext.StoredContext extractAuthenticationContext(BasicHttpRequest request) {
+        HttpHeadersWithAuthenticationContext authenticatedHeaders = unwrapHeadersAuthenticationContext(request);
+        return authenticatedHeaders != null ? authenticatedHeaders.authenticatedContext.get() : null;
+    }
+
+    public static Exception extractAuthenticationException(BasicHttpRequest request) {
+        HttpHeadersWithAuthenticationContext authenticatedHeaders = unwrapHeadersAuthenticationContext(request);
+        return authenticatedHeaders != null ? authenticatedHeaders.authenticationException.get() : null;
+    }
+
+    private static HttpHeadersWithAuthenticationContext unwrapHeadersAuthenticationContext(BasicHttpRequest request) {
+        if (request instanceof Netty4HttpRequest == false) {
+            return null;
+        }
+        if (((Netty4HttpRequest) request).getNettyRequest().headers() instanceof HttpHeadersWithAuthenticationContext == false) {
+            return null;
+        }
+        return ((HttpHeadersWithAuthenticationContext) (((Netty4HttpRequest) request).getNettyRequest().headers()));
+    }
```

Comment:
The methods above are used to extract the attached authentication context from the request (they are used just before the time when the request is to be handled, and the authentication context needs to be instanted).
Unfortunately it relies on `instance of` checks (2 of them), only one of which can be "fixed" because the other uses netty interfaces.